Added CUDA support for torch.orgqr

IvanYashchuk opened the pull request:
**Update:** MAGMA support was dropped from this PR. Only the cuSOLVER path is implemented and it's used exclusively.

**Original PR message:**

This PR adds support for CUDA inputs for `torch.orgqr`.

CUDA implementation is based on both [cuSOLVER](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-orgqr) and MAGMA. cuSOLVER doesn't have a specialized routine for the batched case. While MAGMA doesn't have a specialized GPU native (without CPU sync) `orgqr`. But MAGMA has implemented (and not documented) the batched GPU native version of `larft` function (for small inputs of size <= 32), which together with `larfb` operation form `orgqr` (see the call graph [here at the end of the page](http://www.netlib.org/lapack/explore-html/da/dba/group__double_o_t_h_e_rcomputational_ga14b45f7374dc8654073aa06879c1c459.html)).

So now there are two main codepaths for CUDA inputs (if both MAGMA and cuSOLVER are available):
* if `batchsize > 1` and `tau.shape[-1] <= 32` then MAGMA based function is called
* else [cuSOLVER's `orgqr`](https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-orgqr) is used.

If MAGMA is not available then only cuSOLVER is used and vice versa.

Documentation updates and possibly a new name for this function will be in a follow-up PR.

Ref. https://github.com/pytorch/pytorch/issues/50104

IvanYashchuk committed:
CUDA dispatch for orgqr

IvanYashchuk committed:
Fixed orgqr to return orthogonal matrix; previously only a matrix with orthonormal columns was returned

IvanYashchuk committed:
Added MAGMA batched orgqr

IvanYashchuk committed:
Use skipCUDAIfNoMagmaAndNoCusolver

IvanYashchuk committed:
Added test for batchsize beyond MAGMA's batchlimit

IvanYashchuk commented:
TODO: put here benchmark results.

IvanYashchuk committed:
Revert "Fixed orgqr to return orthogonal matrix; previously only a matrix with orthonormal columns was returned"

This reverts commit c415a0b0c21a303b127090176cf5d8724fb661af.

IvanYashchuk committed:
Fixed magma based apply_orgqr

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -295,6 +295,70 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_lib(const Tensor& self, bool
+  Tensor work = at::empty({lwork}, self.options());
+  auto work_data = work.data_ptr<scalar_t>();
+
+  auto handle = at::cuda::getCurrentCUDASolverDnHandle();
```
This should be inside the for loop.

mruberry commented:
cc @xwang233 @ptrblck please advise on the cuSOLVER performance -- this isn't necessarily a high priority function for us to push on, but if we do want to switch to cuSOLVER exclusively we'll need its performance to be at least comparable to MAGMA. Would the cuSOLVER team be willing to help or work on performance improvements if we identify ops, possibly just high priority ops for now, where they lag behind MAGMA in perf?

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -1196,10 +1196,8 @@ Tensor& orgqr_out(const Tensor& input, const Tensor& tau, Tensor& result) {
-  // therefore we create 'infos' only on CPU for now.
-  // This should be changed if cuSOLVER would be used
-  auto infos = at::empty({std::max<int64_t>(1, batchCount(input))}, input.options().dtype(kInt).device(kCPU));
+  // cuSOLVER is used for CUDA inputs and it requires 'infos' to reside in GPU memory
```
I thought either MAGMA or cuSOLVER could be used?

IvanYashchuk replied:
I should make this comment more clear. Yes, both are used. cuSOLVER requires `infos` to be on GPU, MAGMA path doesn't use it (it can't be used).

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+  * `self` - Tensor with the directions of the elementary reflectors below the diagonal,
+              it will be overwritten with the result
+  * `tau` - Tensor containing the magnitudes of the elementary reflectors
+  * `infos` - Tensor to store MAGMA's error codes (not used here)
```
MAGMA or cuSOLVER?

IvanYashchuk replied:
MAGMA here because this function (and this file) is calling MAGMA.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+template <typename scalar_t>
+inline void apply_orgqr_batched_magma(Tensor& self, const Tensor& tau, Tensor& infos, int64_t n_columns) {
+#ifndef USE_MAGMA
+  TORCH_CHECK(false, "Calling torch.orgqr on a CUDA tensor requires compiling ",
```
I thought cuSOLVER could be used if MAGMA wasn't available?

IvanYashchuk replied:
That's true. But if cuSOLVER is not available (currently it's disabled for cuda version <  10.1.243) then this function will be called and if MAGMA is also not available this error is raised.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+  ALLOCATE_ARRAY(tau_array, scalar_t*, batchsize);
+
+  // Set up the created arrays
+  for (int64_t i = 0; i < batchsize; i++) {
```
type i like batchsize

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+    }
+  }
+
+  // Compute result = (I - V T V^H) Ic,
```
All these comments are great and make the code very readable.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+  self.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).add_(1); // self <- self + Ic: m x n + m x n = m x n
+
+  // infos is not used here, make sure that it's filled with zeros
+  infos.fill_(0);
```
Would you elaborate on why infos is unused in this case and why it needs to be filled with zeros here?

IvanYashchuk replied:
I'll try to improve the comment. `infos` has to be filled with zeros because non-zero content would cause raising an error, alternatively, it could be initialized with zero values. MAGMA path here is more low-level implementation than LAPACK or cuSOLVER and there is no interface to pass `infos` argument to `magmaLarftBatched`.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebra.cu:
```diff
@@ -1893,6 +1951,145 @@ std::tuple<Tensor, Tensor> _triangular_solve_helper_cuda(const Tensor& self, con
+#elif defined(USE_CUSOLVER) && !defined(USE_MAGMA)
+  return orgqr_helper_cuda_lib(result, tau, infos, n_columns); // cusolver
+#else
+  TORCH_CHECK(tau.size(-1) <= 32, "orgqr: tau.shape[-1] larger than 32 is not supported with MAGMA backend. Please use PyTorch built with cuSOLVER support.")
```
We should remember to change the function's signature and rename the second input as tau to clarify these references. (This proposal may be BC-breaking, but it's incredibly unlikely anyone is relying on the current names.)

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -295,6 +295,70 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_lib(const Tensor& self, bool
 }
 
+/*
+  The orgqr function allows reconstruction of an orthogonal (or unitary) matrix Q,
```
Would a single orgqr note that could be referenced in multiple places be easier to maintain?

IvanYashchuk replied:
We don't have proper C++ documentation for the internal functions. So these notes are useful for someone who skims through the source code and I think it's more beneficial to have it right there next to the function.

mruberry commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -295,6 +295,70 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_lib(const Tensor& self, bool
+
+  auto handle = at::cuda::getCurrentCUDASolverDnHandle();
+
+  for (int i = 0; i < batchsize; i++) {
```
type match the loop var with batchsize

mruberry commented on test/test_linalg.py:
```diff
@@ -3942,8 +3942,8 @@ def test_renorm_ps(self, device):
     @skipCPUIfNoLapack
+    @skipCUDAIfNoMagmaAndNoCusolver
     @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
     def test_orgqr(self, device, dtype):
```
How good is this test? Do we need to revise its coverage while we're working on it?

IvanYashchuk replied:
I think this test is as good as it could be because I wrote it 😀
https://github.com/pytorch/pytorch/pull/50502/files#r556989761

It covers 0x0 matrices, 0 batch dimension, normal single matrices, and batched ones.

mruberry commented on test/test_linalg.py:
```diff
@@ -4028,21 +4028,35 @@ def test_orgqr_errors_and_warnings(self, device):
+    @onlyCUDA
+    @skipCUDAIfNoMagmaAndNoCusolver
+    @dtypes(torch.float32)
+    def test_orgqr_many_batches(self, device, dtype):
```
How long does this test take to run?

IvanYashchuk replied:
It looks scary, yes, but it's faster than other tests!
```
3.98s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_complex128
0.40s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_float64
0.33s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_complex128
0.24s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_complex64
0.23s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_float32
0.18s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_errors_and_warnings_cuda
0.14s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_complex64
0.14s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_many_batches_cuda_float32
0.07s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_errors_and_warnings_cpu
0.05s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_float32
0.05s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_float64

```
And when MAGMA is not available (using cuSOLVER then)
```
4.40s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_complex128
0.34s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_float32
0.34s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_float64
0.30s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_complex128
0.29s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_cuda_complex64
0.22s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_errors_and_warnings_cuda
0.19s call     test/test_linalg.py::TestLinalgCUDA::test_orgqr_many_batches_cuda_float32
0.03s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_errors_and_warnings_cpu
0.02s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_complex64
0.01s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_float32
0.01s call     test/test_linalg.py::TestLinalgCPU::test_orgqr_cpu_float64
```

mruberry reviewed (commented):
Hey @IvanYashchuk! This PR looks great. I made a few questions/comments inline. We should be sure to add better support for nondifferentiable functions to OpInfos and create an OpInfo for orgqr, too, when we come back to update its documentation.

xwang233 commented:
@IvanYashchuk thanks for the PR and thanks @mruberry  for the comments. From the benchmark here https://github.com/pytorch/pytorch/pull/51348#issuecomment-769749065, it seems like cusolver implementation only has better performance when batch size is `<= 2`? Well, this would be a similar heuristic like `torch.inverse`. I'll collect data with some repro scripts and send a request to cusolver team.

mruberry commented:
Added you as a reviewer, @xwang233.

IvanYashchuk committed:
Moved handle inside the batching loop

It gives very minor performance gain; it would help better if cuSOLVER's
orgqr wouldn't include synchronous cuda calls.

IvanYashchuk committed:
Use decltype

IvanYashchuk committed:
Updated comments about infos

mruberry commented:
Unfortunately it looks like at least some of the build failures are real and the test failure on Windows is real. The XLA build failure appears to just be a CI issue.

Once the builds are working @heitorschueroff can help you land this, @IvanYashchuk (I'll be out next week).

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-cuda

IvanYashchuk committed:
Fix typo from prev commits

IvanYashchuk commented:
Windows failure is coming from MAGMA internals 😔
```
RuntimeError: CUDA error: unspecified launch failure
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 8
=========     at 0x00000050 in zdisplace_pointers_kernel(float**, float**, int, int, int)
=========     by thread (0,0,0) in block (1,0,0)
=========     Address 0x9dceb3ddc8 is out of bounds
=========     Device Frame:zdisplace_pointers_kernel(float**, float**, int, int, int) (zdisplace_pointers_kernel(float**, float**, int, int, int) : 0x50)
```

xwang233 reviewed (commented):
Thanks for the work. I have left some preliminary comments.

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -297,6 +297,69 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_lib(const Tensor& self, bool
+
+  // LAPACK's requirement
+  TORCH_INTERNAL_ASSERT(m >= n);
+  TORCH_INTERNAL_ASSERT(n >= k);
```
The messages are here https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1270-L1273.
Those internal asserts should never be triggered.

IvanYashchuk commented on aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:
```diff
@@ -297,6 +297,69 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_lib(const Tensor& self, bool
+  at::cuda::solver::orgqr_buffersize<scalar_t>(
+    at::cuda::getCurrentCUDASolverDnHandle(), m, n, k, self_data, lda, tau_data, &lwork);
+  Tensor work = at::empty({lwork}, self.options());
+  auto work_data = work.data_ptr<scalar_t>();
```
Good point, it's not safe. I'll move it inside the loop.

IvanYashchuk commented on aten/src/ATen/cuda/CUDASolver.cpp:
```diff
@@ -335,6 +335,110 @@ void gesvdjBatched<c10::complex<double>>(
 }
 
+template <>
+void orgqr_buffersize<float>(
```
I saw the code for inverse and svd. I think these functions should be just templated wrappers of the underlying library and nothing more exactly mimicking the number of arguments and their order and not hide anything. So that everything is visible what's going from the `apply_*` function.
The output of `*_bufferSize` functions remains the same for the same input sizes, so there is no point in calling it inside the inner batching loop.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-cuda

IvanYashchuk committed:
Updated error message for orgqr test

IvanYashchuk committed:
Fix warning

IvanYashchuk committed:
Moved workspace allocating inside the batching loop because it's safe that way

IvanYashchuk committed:
Removed MAGMA path; it doesn't work for Windows CUDA 10.1

IvanYashchuk commented:
I decided to remove the MAGMA path from this PR as the Windows failure was caused by MAGMA internals. We could try bringing it back in the future possibly with some workarounds to avoid illegal memory accesses.
All CI pass now (the rocm failure seems to be not related?).
@mruberry I think this is ready to be merged.

OpInfo for orgqr and documentation improvements are handled in https://github.com/pytorch/pytorch/pull/52637.

mruberry commented:
Reverted this because it broke ROCm testing, example failure:

```
16:46:59 ======================================================================
16:46:59 ERROR [0.047s]: test_orgqr_cuda_float64 (__main__.TestLinalgCUDA)
16:46:59 ----------------------------------------------------------------------
16:46:59 Traceback (most recent call last):
16:46:59   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 893, in wrapper
16:46:59     method(*args, **kwargs)
16:46:59   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 292, in instantiated_test
16:46:59     raise rte
16:46:59   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 287, in instantiated_test
16:46:59     result = test_fn(self, *args)
16:46:59   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 572, in dep_fn
16:46:59     return fn(slf, device, *args, **kwargs)
16:46:59   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 572, in dep_fn
16:46:59     return fn(slf, device, *args, **kwargs)
16:46:59   File "test_linalg.py", line 4493, in test_orgqr
16:46:59     run_test(shape)
16:46:59   File "test_linalg.py", line 4473, in run_test
16:46:59     actual = torch.orgqr(reflectors, tau)
16:46:59 RuntimeError: Calling torch.orgqr on a CUDA tensor requires compiling PyTorch with cuSOLVER. Please use PyTorch built with cuSOLVER 
```

Unfortunately the ROCm signal was obscured on the PR CI. @IvanYashchuk, looks like we just need some skips?

IvanYashchuk committed:
Added skipCUDAIfRocm

IvanYashchuk commented:
Oh, that's unfortunate.
@mruberry I added the skips now.

mruberry commented:
Great! Let's just wait for the CI.

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-cuda

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-cuda

IvanYashchuk committed:
Merge remote-tracking branch 'upstream/master' into orgqr-cuda

IvanYashchuk commented:
@mruberry all checks have passed. I think it's safe to merge it now.
